File: data_structures/trees/binary_tree/typescript/binaryTreeMinDepth.ts
import BinaryTreeNode from './BinaryTreeNode';
import Queue from '@ds/queue/typescript/Queue';

/**
 * Returns binary tree's minimum depth
 */
export function binaryTreeMinDepthRecursive<T>(root: BinaryTreeNode<T>): number {
    if (!root) {
        return 0;
    }

    const left = binaryTreeMinDepthRecursive(root.left);
    const right = binaryTreeMinDepthRecursive(root.right);

    if (left === 0 || right === 0) {
        return Math.max(left, right) + 1;
    } else {
        return Math.min(left, right) + 1;
    }
}

export function binaryTreeMinDepthIterativeBFS<T>(root: BinaryTreeNode<T>): number {
    if (!root) {
        return 0;
    }

    const queue = new Queue<BinaryTreeNode<T>>();
    queue.enqueue(root);

    let depth = 1;

    while (!queue.isEmpty) {
        const size = queue.length;
        for (let i = 0; i < size; i++) {
            const currNode = queue.dequeue();

            if (currNode.left === null && currNode.right === null) {
                return depth;
            }

            if (currNode.left) {
                queue.enqueue(currNode.left);
            }

            if (currNode.right) {
                queue.enqueue(currNode.right);
            }
        }

        depth++;
    }

    return depth;
}

File: data_structures/trees/binary_tree/python/binary_tree_min_depth.py
from typing import TypeVar

from data_structures.queue.pyhton.queue import Queue
from data_structures.trees.binary_tree.python.binary_tree_node import BinaryTreeNode

T = TypeVar('T')


def binary_tree_min_depth_recursive(root: BinaryTreeNode[T]) -> int:
    """
    Returns min depth of binary tree
    """
    if root is None:
        return 0

    left = binary_tree_min_depth_recursive(root.left)
    right = binary_tree_min_depth_recursive(root.right)

    if left == 0 or right == 0:
        return max(left, right) + 1
    else:
        return min(left, right) + 1


def binary_tree_min_depth_iterative_bfs(root: BinaryTreeNode[T]) -> int:
    if not root:
        return 0

    queue: Queue[BinaryTreeNode[T]] = Queue()
    queue.enqueue(root)
    depth = 1

    while not queue.is_empty:
        size = queue.is_empty + 1
        for i in range(size):
            curr_node: BinaryTreeNode[T] = queue.dequeue()

            if not curr_node.left and not curr_node.right:
                return depth

            if curr_node.left:
                queue.enqueue(curr_node.left)

            if curr_node.right:
                queue.enqueue(curr_node.right)

        depth += 1

    return depth

